lib/puppet/functions.rb

```diff
@@ -260,25 +284,51 @@ def initialize(dispatcher, type_parser, all_callables)
       @dispatcher = dispatcher
     end
 
-    # Defines a positional parameter with type and name
+    # Defines a required positional parameter with a _type_ and a _name_.
     #
     # @param type [String] The type specification for the parameter.
-    # @param name [String] The name of the parameter. This is primarily used
+    # @param name [Symbol] The name of the parameter. This is primarily used
     #   for error message output and does not have to match the name of the
     #   parameter on the implementation method.
     # @return [Void]
     #
     # @api public
     def param(type, name)
-      raise ArgumentError, 'Parameters cannot be added after a block_param' unless @block_type.nil?
-      if type.is_a?(String)
-        @types << type
-        @names << name
-        # mark what should be picked for this position when dispatching
-        @weaving << @names.size()-1
-      else
-        raise ArgumentError, "Type signature argument must be a String reference to a Puppet Data Type. Got #{type.class}"
-      end
+      internal_param(type, name)
+      raise ArgumentError, 'A required parameter cannot be added after an optional parameter' if @min != @max
+      @min += 1
+      @max += 1
+    end
+    alias required_param param
+
+    # Defines an optional positional parameter with a _type_ and a _name_.
+    # May not be used on a parameter that is followed by a required parameter
+    #
+    # @param type [String] The type specification for the parameter.
+    # @param name [Symbol] The name of the parameter. This is primarily used
+    #   for error message output and does not have to match the name of the
+    #   parameter on the implementation method.
```

Comment:
"and does not have to match an implementation method parameter" (I don't think of parameters as being "on" a method)